turbojpeg.py

```diff
@@ -227,6 +229,76 @@ def encode(self, img_array, quality=85, pixel_format=TJPF_BGR, jpeg_subsample=TJ
+    @scaling_factor     optional image scaling
+    @quality            optional image quality to be reencoded with
+    @flags
+    """
+    def scale_with_quality(self, jpeg_buf, pixel_format=TJPF_BGR, jpeg_subsample=TJSAMP_420,
+        scaling_factor=None, quality=85, flags=0):
+        """decompresstoYUV with scale factor, recompresstoYUV with quality factor"""
+        handle = self.__init_decompress()
+        try:
+            if scaling_factor is not None and \
+                scaling_factor not in self.__scaling_factors:
+                raise ValueError('supported scaling factors are ' +
+                    str(self.__scaling_factors))
+            pixel_size = [3, 3, 4, 4, 4, 4, 1, 4, 4, 4, 4, 4]
+            width = c_int()
+            height = c_int()
+            jpeg_subsample = c_int()
+            jpeg_colorspace = c_int()
+            jpeg_array = np.frombuffer(jpeg_buf, dtype=np.uint8)
+            src_addr = self.__getaddr(jpeg_array)
+            status = self.__decompress_header(
+                handle, src_addr, jpeg_array.size, byref(width), byref(height),
+                byref(jpeg_subsample), byref(jpeg_colorspace))
+            if status != 0:
+                self.__report_error(handle)
+            scaled_width = width.value
+            scaled_height = height.value
+            if scaling_factor is not None:
+                def get_scaled_value(dim, num, denom):
+                    return (dim * num + denom - 1) // denom
+                scaled_width = get_scaled_value(
+                    scaled_width, scaling_factor[0], scaling_factor[1])
+                scaled_height = get_scaled_value(
+                    scaled_height, scaling_factor[0], scaling_factor[1])
+            img_array = np.empty(
+                [scaled_height, scaled_width, pixel_size[pixel_format]],
+                dtype=np.uint8)
+            dest_addr = self.__getaddr(img_array)
+            #status = self.__decompress(
+            #    handle, src_addr, jpeg_array.size, dest_addr, scaled_width,
+            #    0, scaled_height, pixel_format, flags)
+            status = self.__decompressToYUV2(
+                handle, src_addr, jpeg_array.size, dest_addr, scaled_width, 4, scaled_height, flags)
+            if status != 0:
+                self.__report_error(handle)
+                return
+            #return img_array
+            self.__destroy(handle)
+
+            handle = self.__init_compress()
+            jpeg_buf = c_void_p()
+            jpeg_size = c_ulong()
+            height, width, _ = img_array.shape
+            src_addr = dest_addr
+            status = self.__compressFromYUV(
+                handle, src_addr, width, 4, height, jpeg_subsample, byref(jpeg_buf),
+                byref(jpeg_size), quality, flags)
+            if status != 0:
+                self.__report_error(handle)
+            dest_buf = create_string_buffer(jpeg_size.value)
+            memmove(dest_buf, jpeg_buf.value, jpeg_size.value)
```

Comment:
Would like to be able to eventually get rid of the double allocation and copy.  But that can wait until later.